src/test/java/com/aws/greengrass/certificatemanager/certificate/CISShadowMonitorTest.java

```diff
@@ -49,172 +77,507 @@
-        ArgumentCaptor<GetShadowSubscriptionRequest> getSubscribeRequest = ArgumentCaptor.forClass(
-                GetShadowSubscriptionRequest.class);
-        ArgumentCaptor<GetShadowRequest> publishRequest = ArgumentCaptor.forClass(GetShadowRequest.class);
+    @Test
+    void GIVEN_CISShadowMonitor_WHEN_cis_shadow_changed_multiple_times_THEN_new_cert_generated_multiple_times() throws Exception {
+        int numShadowChanges = 3;
+
+        CountDownLatch shadowDeltaUpdated = whenShadowDeltaUpdated(numShadowChanges);
+
+        cisShadowMonitor.startMonitor();
+        cisShadowMonitor.addToMonitor(certificateGenerator);
 
-        verify(mockShadowClient, times(1)).SubscribeToShadowDeltaUpdatedEvents(
-                updateSubscribeRequest.capture(), eq(QualityOfService.AT_LEAST_ONCE), any(),
-                any());
-        assertThat(updateSubscribeRequest.getValue().thingName, is(SHADOW_NAME));
-        verify(mockShadowClient, times(1)).SubscribeToGetShadowAccepted(
-                getSubscribeRequest.capture(), eq(QualityOfService.AT_LEAST_ONCE), any(),
-                any());
-        assertThat(getSubscribeRequest.getValue().thingName, is(SHADOW_NAME));
+        IntStream.range(0, numShadowChanges).forEach(i -> {
+            publishDesiredShadowState(Utils.immutableMap("newState", i));
+        });
 
-        verify(mockShadowClient, times(1)).PublishGetShadow(publishRequest.capture(),
-                eq(QualityOfService.AT_LEAST_ONCE));
-        assertThat(publishRequest.getValue().thingName, is(SHADOW_NAME));
+        assertTrue(shadowDeltaUpdated.await(5L, TimeUnit.SECONDS));
+        verify(certificateGenerator, times(numShadowChanges)).generateCertificate(any(), any());
     }
 
     @Test
-    public void GIVEN_CISShadowMonitor_WHEN_update_delta_version_THEN_cert_generated() throws KeyStoreException {
-        CISShadowMonitor cisShadowMonitor = new CISShadowMonitor(mockMqttClient, mockConnection, mockShadowClient,
-                mockExecutor, SHADOW_NAME, mockConnectivityInfoProvider);
-        when(mockShadowClient.SubscribeToShadowDeltaUpdatedEvents(any(), any(), any(), any()))
-                .thenReturn(CompletableFuture.completedFuture(0));
-        when(mockShadowClient.SubscribeToGetShadowAccepted(any(), any(), any(), any()))
-                .thenReturn(CompletableFuture.completedFuture(0));
+    void GIVEN_CISShadowMonitor_WHEN_cis_shadow_changed_duplicate_THEN_only_one_new_cert_generated() throws Exception {
+        CountDownLatch shadowDeltaUpdated = whenShadowDeltaUpdated(2);
+
         cisShadowMonitor.startMonitor();
+        cisShadowMonitor.addToMonitor(certificateGenerator);
+
+        shadowClient.withDuplicatePublishing(true);
+
+        publishDesiredShadowState(Utils.immutableMap("newState", 1));
+
+        assertTrue(shadowDeltaUpdated.await(5L, TimeUnit.SECONDS));
+        verify(certificateGenerator).generateCertificate(any(), any());
```

Comment:
verify generateCertificate() is called only _once_?

Reply:
Yep, duplicate calls to the shadow delta handler are idempotent, we'll only generate the cert the first time

Reply:
I mean we should explicitly verify here that `generateCertificate()` is called only once:
```
verify(certificateGenerator, times(1)).generateCertificate(any(), any());
```

Reply:
Ahh okay.  I personally think `times(1)` is verbose, but looks like it's widely used in the codebase, will adjust :)

Reply:
There is also `once` which is a synonym.

Reply:
done